src/gsi/gsi_rfv3io_mod.f90

```diff
@@ -1902,40 +1912,24 @@ subroutine gsi_fv3ncdf2d_read(fv3filenamegin,it,ges_z,ges_t2m,ges_q2m)
              write(*,*) "wrong dimension number ndim =",ndim
              call stop2(119)
           endif
-          if(allocated(dim_id    )) deallocate(dim_id    )
-          allocate(dim_id(ndim))
           if(fv3_io_layout_y > 1) then
              do nio=0,fv3_io_layout_y-1
-               iret=nf90_inquire_variable(gfile_loc_layout(nio),i,dimids=dim_id)
-               if(allocated(sfc       )) deallocate(sfc       )
-               if(dim(dim_id(1)) == nx .and. dim(dim_id(2))==ny_layout_len(nio)) then
-                  if(ndim >=3) then
-                     allocate(sfc(dim(dim_id(1)),dim(dim_id(2)),dim(dim_id(3))))
-                     iret=nf90_get_var(gfile_loc_layout(nio),i,sfc)
-                  else if (ndim == 2) then
-                     allocate(sfc(dim(dim_id(1)),dim(dim_id(2)),1))
-                     iret=nf90_get_var(gfile_loc_layout(nio),i,sfc(:,:,1))
-                  endif
-               else
-                  write(*,*) "Mismatch dimension in surfacei reading:",nx,ny_layout_len(nio),dim(dim_id(1)),dim(dim_id(2))
-                  call stop2(119)
-               endif
-               sfc_fulldomain(:,ny_layout_b(nio):ny_layout_e(nio))=sfc(:,:,1)
+                if(allocated(sfc       )) deallocate(sfc       )
+                allocate(sfc(nx,ny_layout_len(nio),1))
+                if(ndim >=3) then
+                   iret=nf90_get_var(gfile_loc_layout(nio),i,sfc)
+                else if (ndim == 2) then
+                   iret=nf90_get_var(gfile_loc_layout(nio),i,sfc(:,:,1))
+                endif
+                sfc_fulldomain(:,ny_layout_b(nio):ny_layout_e(nio))=sfc(:,:,1)
              enddo
           else
-             iret=nf90_inquire_variable(gfile_loc,i,dimids=dim_id)
              if(allocated(sfc       )) deallocate(sfc       )
-             if(dim(dim_id(1)) == nx .and. dim(dim_id(2))==ny) then
-                if(ndim >=3) then  !the block of 10 lines is compied from GSL gsi.
-                   allocate(sfc(dim(dim_id(1)),dim(dim_id(2)),dim(dim_id(3))))
-                   iret=nf90_get_var(gfile_loc,i,sfc)
-                else if (ndim == 2) then
-                   allocate(sfc(dim(dim_id(1)),dim(dim_id(2)),1))
-                   iret=nf90_get_var(gfile_loc,i,sfc(:,:,1))
-                endif
-             else
-                write(*,*) "Mismatch dimension in surfacei reading:",nx,ny,dim(dim_id(1)),dim(dim_id(2))
-                call stop2(119)
+             allocate(sfc(nx,ny,1))
```

Comment:
A little confusing to me.  When ndim==3, in the original code, the sfc be as original 1930, now, it is always as the current line 1928, with the 3th dimension of 1.  What is supposed to happen when ndim ==3 , when sfc in the files are of vertical dimension >1 ?

Reply:
The one version of FV3 model write out surface fields as 3 dimension fields. The third dimension of this array is always 1 but it fit to the requirement of the dimension number when nf90_get_var is used to read 3D or 2D array.